packages/babel-plugin-transform-class-properties/src/index.js

```diff
@@ -9,21 +9,212 @@ export default function({ types: t }) {
         this.push(path.parentPath);
       }
     },
+
+    ClassBody(path) {
+      path.skip();
+    },
+
+    Function(path) {
+      if (path.isArrowFunctionExpression()) return;
+      path.skip();
+    },
   };
 
-  const referenceVisitor = {
+  const collisionVisitor = {
     TypeAnnotation(path) {
       path.skip();
     },
+
     ReferencedIdentifier(path) {
       if (this.scope.hasOwnBinding(path.node.name)) {
         this.collision = true;
-        path.skip();
+        path.stop();
+      }
+    },
+  };
+
+  const remapThisVisitor = {
+    ThisExpression(path) {
+      path.replaceWith(this.thisRef);
+    },
+
+    Function(path) {
+      if (path.isArrowFunctionExpression()) return;
+      path.skip();
+    },
+  };
+
+  function collectPropertiesVisitor(body, propNames) {
+    for (const path of body.get("body")) {
+      if (path.isClassMethod({ kind: "constructor" })) {
+        propNames.constructor = path;
+        continue;
+      }
+
+      if (!path.isProperty()) continue;
+
+      const { key, computed, static: isStatic } = path.node;
+      propNames[isStatic ? "staticProps" : "instanceProps"].push(path);
+
+      if (computed) continue;
+
+      const isPrivate = path.isClassPrivateProperty();
+      const name = isPrivate
+        ? key.id.name
+        : t.isIdentifier(key) ? key.name : key.value;
+      const seen =
+        propNames[
+          isPrivate
+            ? "privateProps"
+            : isStatic ? "publicStaticProps" : "publicProps"
+        ];
+
+      if (seen[name]) {
+        throw path.buildCodeFrameError("duplicate class field");
+      }
+      seen[name] = true;
+    }
+  }
+
+  const staticErrorVisitor = {
+    ClassBody(path) {
+      path.skip();
+    },
+
+    ClassProperty(path) {
+      const { computed, key, static: isStatic } = path.node;
+      if (computed || !isStatic) return;
+
+      const name = t.isIdentifier(key) ? key.name : key.value;
+      if (name === "prototype") {
+        throw path.buildCodeFrameError("illegal static class field");
       }
     },
+
+    PrivateName(path) {
+      const { parentPath, node } = path;
+      if (parentPath.isMemberExpression({ property: node, computed: false })) {
+        if (!this.privateProps[node.id.name]) {
+          throw path.buildCodeFrameError("unknown private property");
+        }
+
+        return;
+      }
+
+      if (parentPath.isClassPrivateProperty({ key: node })) return;
+
+      throw path.buildCodeFrameError(
+        `illegal syntax. Did you mean \`this.#${node.id.name}\`?`,
```

Comment:
The error message here could refer to the fact that this is proposed as a follow-on, at https://github.com/littledan/proposal-private-shorthand . The committee has specifically discussed that it'd be great if Babel error messages for shorthand property usage to refer to this. It would be fine to make as a separate transform and include in the Stage 2 preset. cc @wycats